src/netconf_acm.c

```diff
@@ -0,0 +1,1862 @@
+                    str = lyd_get_value(node);
+                    free(rule->target);
+                    if (!strcmp(str, "*")) {
+                        rule->target = NULL;
+                    } else {
+                        rule->target = strdup(str);
+                    }
+                    if (!strcmp(node->schema->name, "rpc-name")) {
+                        rule->target_type = SR_NACM_TARGET_RPC;
+                    } else if (!strcmp(node->schema->name, "notification-name")) {
+                        rule->target_type = SR_NACM_TARGET_NOTIF;
+                    } else {
+                        assert(!strcmp(node->schema->name, "path"));
+                        rule->target_type = SR_NACM_TARGET_DATA;
+                    }
+                }
+            } else if (!strcmp(node->schema->name, "access-operations")) {
+                str = lyd_get_value(node);
+                rule->operations = 0;
+                if (!strcmp(str, "*")) {
+                    rule->operations = SR_NACM_OP_ALL;
+                } else {
+                    if (strstr(str, "create")) {
+                        rule->operations |= SR_NACM_OP_CREATE;
+                    }
+                    if (strstr(str, "read")) {
+                        rule->operations |= SR_NACM_OP_READ;
+                    }
+                    if (strstr(str, "update")) {
+                        rule->operations |= SR_NACM_OP_UPDATE;
+                    }
+                    if (strstr(str, "delete")) {
+                        rule->operations |= SR_NACM_OP_DELETE;
+                    }
+                    if (strstr(str, "exec")) {
+                        rule->operations |= SR_NACM_OP_EXEC;
+                    }
+                }
+            } else if (!strcmp(node->schema->name, "action")) {
+                if (!strcmp(lyd_get_value(node), "permit")) {
+                    rule->action_deny = 0;
+                } else {
+                    rule->action_deny = 1;
+                }
+            } else if (!strcmp(node->schema->name, "comment")) {
+                if (op == SR_OP_DELETED) {
+                    free(rule->comment);
+                    rule->comment = NULL;
+                } else {
+                    assert((op == SR_OP_MODIFIED) || (op == SR_OP_CREATED));
+                    free(rule->comment);
+                    rule->comment = strdup(lyd_get_value(node));
+                }
+            }
+        }
+    }
+
+    /* NACM UNLOCK */
+    pthread_mutex_unlock(&nacm.lock);
+
+    sr_free_change_iter(iter);
+    if (rc != SR_ERR_NOT_FOUND) {
+        sr_session_set_error_message(session, "Getting next change failed (%s).", sr_strerror(rc));
+        return rc;
+    }
+
+    return SR_ERR_OK;
+}
+
+#define SR_CONFIG_SUBSCR(session, mod_name, xpath, opts, cb) \
+    rc = sr_module_change_subscribe(session, mod_name, xpath, cb, NULL, opts, \
+            SR_SUBSCR_DONE_ONLY | SR_SUBSCR_ENABLED, sub); \
```

Comment:
Instead of `opts` there should be `0`, it is the priority and instead of the following fixed flags there should be `opts`. Also, I would like `sub` to be part of the macro, this is dangerous.

Reply:
Okay, I have fixed the error and sub is now part of the macro. However, aren't the fixed flags needed for proper functionality of the callbacks? As in, they assume these flags will be used. Or maybe your point is that I should OR those flags at the location the marco is used?

Reply:
I sort-of stopped thinking when I say the parameter problem. Yes, the 2 flags need to be set and `SR_SUBSCR_PASSIVE` must not be. The only flag that should actually be supported is `SR_SUBSCR_NO_THREAD` so maybe put that in doxygen and check the `opts` param accordingly returning inval arg is some other options are set.

Reply:
Yeah, that makes sense. `SR_CHECK_ARG_APIRET` macro returns an `int`, I wonder if all the functions in the API should return an int? What would happen with the errors then

Reply:
Oh, right, I have missed that. Yes, so these functions should all return `int`, `error_info` is used only internally and set to `session` when passing API. But `session` is not passed to these functions even though `user` is. One solution would be to pass `session` instead of `user` but then the NACM user would need to be stored in the `session` like I was suggesting before. I do not really have a problem with this.